src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs

```diff
@@ -303,33 +298,28 @@ private DecisionTree AddByType(DecisionTree.ByType byType, TypeSymbol type, Deci
                     }
                 }
             }
-            foreach (var kvp in byType.TypeAndDecision)
+
+            // if the last type is the type we need, add to it
+            DecisionTree result = null;
+            if (byType.TypeAndDecision.Count != 0)
             {
-                var MatchedType = kvp.Key;
-                var Decision = kvp.Value;
-                // See if matching Type matches this value
-                switch (ExpressionOfTypeMatchesPatternType(type, MatchedType, ref _useSiteDiagnostics))
+                var lastTypeAndDecision = byType.TypeAndDecision.Last();
+                if (lastTypeAndDecision.Key.TupleUnderlyingTypeOrSelf() == type.TupleUnderlyingTypeOrSelf())
```

Comment:
Please add targeted tests at least for dynamic and tuple name mismatch (both not limited to the top level).